sdk/servicebus/Azure.Messaging.ServiceBus/tests/Plugins/PluginLiveTests.cs

```diff
@@ -314,5 +339,45 @@ public override ValueTask AfterMessageReceiveAsync(ServiceBusReceivedMessage mes
                 throw new NotImplementedException();
             }
         }
+
+        internal class ProcessPlugin : ServiceBusPlugin
```

Comment:
This is an example of a plugin a customer could write if we were to expose the Process* methods. Note that currently, the pipeline implementation doesn't actually send the message at the end of the pipeline, but if we were to expose these methods, we would probably update the implementation to put the send at the end of the pipeline to enable potential telemetry scenarios.

Reply:
Because sending is in batches I don't think sending at the end of the pipeline works well with processing one message at a time.

Reply:
I realize this is just an example, but it's important to note that you also would break the `TryAdd` semantics of adding messages to a batch by calling `Send`.   Essentially, you would only be able to use it in a scenario where you were intentionally sending single message batches or you had a `Send` operation that was buffered.

Reply:
The value of the stack based approach is diminished if we don't have the send at the end of the pipeline. You are right that since we need to loop over all the messages and run the full pipeline for each message, we couldn't just send at the end of each pipeline though.

To Jesse's point, yes this wouldn't be possible for batched sending, thanks for calling that out.

Reply:
2 comments:
1. Would it make sense to name these `ProcessSendMessageAsync` and `ProcessReceiveMessageAsync` as per our recent `Message` naming changes for everything else?
2. Would it make sense to add `ProcessSendBatchAsync` and `ProcessReceiveBatchAsync` for the batch scenario?